File: WEEK_3_C/Q7.c
#include <stdio.h>
#include <stdlib.h>

struct Node{ int data; struct Node* next; };

struct Node* merge(struct Node* a, struct Node* b){
    if(!a) return b; if(!b) return a;
    if(a->data<b->data){ a->next=merge(a->next,b); return a; }
    else{ b->next=merge(a,b->next); return b; }
}

void print(struct Node* h){ while(h){ printf("%d ",h->data); h=h->next; } printf("\n"); }

int main(){
    struct Node *a=NULL,*b=NULL,*t;
    for(int i=3;i>=1;i--){ t=(struct Node*)malloc(sizeof(struct Node)); t->data=i*2; t->next=a; a=t; }
    for(int i=3;i>=1;i--){ t=(struct Node*)malloc(sizeof(struct Node)); t->data=i*2-1; t->next=b; b=t; }
    struct Node* head=merge(a,b);
    print(head);
    return 0;
}

File: WEEK_3_Python/Q7.py
class Node:
    def __init__(self, data):
        self.data, self.next = data, None

def print_list(head):
    cur = head
    while cur:
        print(cur.data, end=" ")
        cur = cur.next
    print()

def remove_loop(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            slow = head
            while slow.next != fast.next:
                slow = slow.next
                fast = fast.next
            fast.next = None
            return head
    return head

# Example
head = Node(1); head.next = Node(2); head.next.next = Node(3)
head.next.next.next = head.next   # loop
head = remove_loop(head)
print_list(head)   # 1 2 3
